src/use_case/take_quiz/TakeQuizInteractor.java

```diff
@@ -15,25 +18,90 @@
+    ArrayList<Song> masterlist = quizDao.getSongList();
+
+    //////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    // STEP 3: Create Quiz Object
+    Quiz quiz = new Quiz();
+    // IMPORTANT NOTE: Below quiz.setNumQuestions method doesn't work unless...
+    // you put {} around it, it's because it returns void.
+    // Find a way to set limit of input to be <= 50.
+    {quiz.setNumQuestions(50);}
+
+    /////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    // STEP 4: Helper method for GENERATING 2 RANDOM SONGS FROM MASTERLIST
+    public ArrayList<Song> generate2RandomSongs(){
        // Creating a Random object
         Random rand = new Random();
 
         // Creating an empty ArrayList to store 2 random songs
         ArrayList<Song> finalSongs = new ArrayList<Song>();
 
         // Getting a random list index 1 and 2
-        int ranIndex1 = rand.nextInt(this.songList.size());
-        int ranIndex2 = rand.nextInt(this.songList.size());
-        // If indexes are equal, we choose a new random index
-        while(ranIndex1 == ranIndex2){
-            ranIndex2 = rand.nextInt(this.songList.size());
+        int size = masterlist.size();
+        int ranIndex1 = rand.nextInt(size);
+        int ranIndex2 = rand.nextInt(size);
+
+        Song ranSong1 = masterlist.get(ranIndex1);
+        Song ranSong2 = masterlist.get(ranIndex2);
+
+        // If indexes are equal OR songs have same popularity, we choose a new random index/song
+        while((ranIndex1 == ranIndex2) || (ranSong1.getPopularity() == ranSong2.getPopularity())){
+            ranIndex2 = rand.nextInt(masterlist.size());
             }
-        Song ranSong1 = this.songList.get(ranIndex1);
-        Song ranSong2 = this.songList.get(ranIndex2);
 
+        Song finalRanSong2 = masterlist.get(ranIndex2);
+
+        // Storing both songs in empty ArrayList
         finalSongs.add(ranSong1);
-        finalSongs.add(ranSong2);
+        finalSongs.add(finalRanSong2);
     return finalSongs;
     }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    // STEP 5: Generate all questions and store in Quiz.questions attribute
+    public void generateAllQuestions(){
+        int i = 0;
+        ArrayList<Question> questionList = new ArrayList<Question>();
+        // Loop through as many times as num_questions
+        while (i <= quiz.getNumQuestions()) {
+            // Create a Question object
+            Question question = new Question();
+            // Generating 2 random songs and setting those as Question.song1 and Question.song2 attributes
+            question.setSongs(this.generate2RandomSongs());
+            // Setting Question attributes Question.pointsAwarded and Question.answer
+            question.setAnswer();
+            question.setPointsAwarded();
```

Comment:
question.setAnswer() and question.setPointsAwarded can both just be done in the constructor for Question class, it might make things look a bit cleaner. If we just put the code for question.setSongs() in the constructor and then calculate the answer and points awarded there that should work.